src/core/shape_inference/include/reshape_shape_inference.hpp

```diff
@@ -0,0 +1,283 @@
+        auto& out_interval = out.get_interval();
+        if (product_out.get_min_length() != 0 && out_interval != Interval{} && product_out != TDim{}) {
+            out_interval.set_max_val(out_interval.get_max_val() / product_out.get_min_length());
+        }
+        if (product_out.get_max_length() != 0) {
+            out_interval.set_min_val(ceil_div(out_interval.get_min_val(), product_out.get_interval().get_max_val()));
+        }
+    }
+    return out;
+}
+
+// resolve minus one dimension for static dimension
+template <class TDim,
+          typename std::enable_if<!std::is_same<typename std::decay<TDim>::type, Dimension>::value>::type* = nullptr>
+TDim resolve_minus_one_dim(const Product<TDim>& product_in, const TDim& product_out) {
+    return product_in.total() / product_out.get_length();
+}
+
+/**
+ * @brief Get the pattern and minus one idx from input bounds.
+ *
+ * @param op      Pointer to reshape node.
+ * @param bounds  Vector of reshape pattern bounds.
+ *
+ * @return Pair which got bounds converted to shape and `minus_one` index in pattern (-1 if not found).
+ */
+template <class TShape>
+std::pair<TShape, int64_t> get_pattern_and_minus_one_idx(const Node* const op,
+                                                         const std::vector<std::pair<int64_t, int64_t>>& bounds) {
+    using namespace ov::util;
+    const auto minus_one_bound = std::make_pair(dim::inf_bound, dim::inf_bound);
+
+    auto result = std::make_pair(TShape{}, dim::inf_bound);
+    auto& shape = std::get<0>(result);
+    shape.reserve(bounds.size());
+
+    auto& minus_one_idx = std::get<1>(result);
+    auto bounds_iter = bounds.begin();
+
+    for (size_t i = 0; i < bounds.size(); ++i, ++bounds_iter) {
+        if (*bounds_iter == minus_one_bound) {
+            NODE_VALIDATION_CHECK(op, minus_one_idx == dim::inf_bound, "More than one dimension has size of -1");
+            minus_one_idx = static_cast<int64_t>(i);
+        }
+        NODE_VALIDATION_CHECK(op, *bounds_iter >= minus_one_bound, "Dim size cannot be less than -1");
+        shape.emplace_back(bounds_iter->first, bounds_iter->second);
+    }
+
+    return result;
+}
+
+/**
+ * @brief Set the pattern labels on pattern shape if this input is labeled.
+ *
+ * @param op     Pointer to reshape node.
+ * @param shape  Pointer to shape for labels set.
+ */
+template <class TShape, typename std::enable_if<std::is_same<TShape, PartialShape>::value>::type* = nullptr>
+void set_pattern_labels(const Node* const op, TShape& shape) {
+    if (op->get_input_size() > 0) {
+        auto labels = op->get_input_source_output(1).get_tensor().get_value_label();
+
+        if (!labels.empty()) {
+            auto label_iter = labels.begin();
+            for (auto& d : shape) {
+                if (*label_iter != no_label) {
+                    DimensionTracker::set_label(d, *label_iter);
+                }
+                ++label_iter;
+            }
+        }
+    }
+}
```

Comment:
What if value 0 or -1 has a label -- we don't want this label to override labels that we deduced from input shape

Reply:
What I tested the label should not be overwritten, but if you got some special cases I will test them and check if updated implementation handles it correctly.

Reply:
Pseudo code:
```
input = Parameter([1, 1, 100]); // label Dimension 100 with label 1

pattern = Parameter([1]);
pattern.output(0).get_tensor().set_value_labels({2});
tensor_m1 = Tensor({-1});
pattern.output(0).get_tensor().set_lower_bound(tensor_m1);
pattern.output(0).get_tensor().set_upper_bound(tensor_m1);

test = Reshape(input, pattern, false);
ASSERT(DimensionTracker::get_label(test.get_output_partial_shape(0)[0]) == 2);
```

You can achieve the same with zero, but the outcome of labels resolving would depend on special_zero attribute of Reshape operation.

Reply:
Updated minus one dimension calculation and label propagation.
See test groups with for details:
- `reshape_label_propagation_minus_one_no_special_zero_case_X`
- `reshape_label_propagation_minus_one_special_zero_case_X`
- `reshape_tricky_label_propagation_for_auto_batch_case_X`

The tests pass also on master version except two cases `reshape_label_propagation_minus_one_special_zero_case_6` and `reshape_tricky_label_propagation_for_auto_batch_case_4`. The difference is that on master the label is not propagated but it can be.